Better handling of json_encode failure, especially in debug mode

dereuromark opened the pull request:
Resolves https://github.com/cakephp/cakephp/issues/8130

Currently you get a white screen of death in debug mode and if json_encode doesnt like your stacktrace.
This removes the stacktrace and tries again.
As last resort at least a trigger_error() call gives the developer a hint what is going on.

I needed quite a while to track the issue down to the root cause, and luckily had a debugger to step through, otherwise it would have been close to impossible..

ADmad commented:
Shouldn't error relating to json be in JsonView?  This is the parent class for xmlview and jsonview.

dereuromark commented:
Good point.

dereuromark commented:
I adjusted it to be generically aware of boolean false returned and throw a more generic but clear error message into the output/logs.

Commit:
Better handling of json_encode failure, especially in debug mode.
also doc block is lying without the cast.

markstory commented on src/View/SerializedView.php:
```diff
@@ -84,8 +85,23 @@ public function render($view = null, $layout = null)
-        } elseif ($view !== false && $this->_getViewFileName($view)) {
+            $result = $this->_serialize($serialize);
+            if ($result === false) {
+                if (Configure::read('debug') && in_array('trace', $serialize)) {
```
What if serialize is `true`?

markstory replied:
Instead of trying to hack out the trace, could we just trigger the error and let people know their data is not going to work?

dereuromark replied:
Well, it happened in debug mode without a good reason (There was nothing in the stack trace that should have let this happen).
So as a result you would not be able to work locally anymore :/

markstory replied:
I suspect there was something unserializable in the trace. I wasn't able to reproduce this error in bare-bones application I have setup.

jippi replied:
:+1: for not trying to be clever, just fail quick and let the developer handle it

dereuromark replied:
There is on multiple systems, same thing happens on jenkins and a debian system.
So there is a good chance this will happen to more people.

markstory replied:
Given that its only on some traces, wouldn't you agree that the error is caused by something in tfe traceback context?

dereuromark replied:
Yes, sure. But it is out of the hands of the developer in that case, and should at least try to avoid the exeption if possible (the other option of getting exceptions in your face and not knowing how to fix them is not really feasible). This piece is only relevant in debug mode, so non-debug mode is not affected by this change.

markstory replied:
Sniffing around for blacklisted keys feels like it could end us up in trouble longer term. This reminds me of when we hid all the connection related keys in debug() and had to revert many of them as they were hiding useful data.

markstory commented on src/View/SerializedView.php:
```diff
@@ -84,8 +85,23 @@ public function render($view = null, $layout = null)
+                        return $result;
+                    }
+                }
+                trigger_error('Serialization of View data failed.');
```
Normally we would throw an exception in these cases.

dereuromark replied:
This would not be too BC for those cases where it happens and people live with it currently.
But I would agree.

markstory replied:
New exceptions are not 'bc breaking' according to our policies.

jippi replied:
:+1: for exceptions

dereuromark replied:
What exception would that be?

Commit:
Use Exception. also fixes wrong deprecation exceptions.

dereuromark committed:
Fix CS.

dereuromark commented:
Should be good to go.

markstory committed:
Remove traces from serialized content.

Trace data cannot always be safely serialized. Because of this we'll
exclude the trace from serialized views. While this might be considered
a breaking change not being able to get error pages in dev mode is very
annoying.

markstory commented:
@dereuromark I've pushed up the changes I suggested in IRC, does this work for you?

ADmad commented on src/Error/ExceptionRenderer.php:
```diff
@@ -184,7 +184,6 @@ public function render()
                 'format' => 'array',
                 'args' => false
             ]);
-            $viewVars['_serialize'][] = 'trace';
```
This means we won't have traces anymore for json error pages?

dereuromark replied:
No, that means we still DO by default, but if they contain un-renderable content that it wont fail hard, but gracefully just dont display the trace in that case (as it would result in a white screen of death).

dereuromark replied:
Actually, I think that line should not have to be changed..

dereuromark replied:
Ah, it wasn't my change, that's why I was confused.
Well, not having them at all (and logging it instead?) is also OK for me, but it wouldn't have to be this way with (my) previous code. In many cases it works fine, but in those it doesn't my solution is a more graceful fallback than not dislaying them at all - maybe.

markstory replied:
@ADmad Yes there wouldn't be traces anymore. I thought this was better than sometimes not getting an error response at all.

dereuromark replied:
@markstory With my approach you always get a response, the downsides are more theoretical and any conflicts very unlikely and only relevant for debug mode anway :)

markstory replied:
You still get a response with this change as well. Just not the stack trace. The stacktrace will still be in the logs.

dereuromark replied:
Ok :) logs sound good

ionas replied:
:+1: Sane.

ionas replied:
On a note, I just wanted to add info about the new behavior of stack traces only going to debug.log/error.log when using XmlView/JsonView to have a way for the developer to have good exception reporting...
... and double checked the logs and don't see anything. If debug is on, http://devicesdbservices.dev/api/search/manufacturers will show Unauthorized while http://devicesdbservices.dev/api/search/manufacturers.xml and http://devicesdbservices.dev/api/search/manufacturers.json won't write anything to debug.log.
Shouldn't these requests do that?

markstory replied:
The error logs are always written to error.log. However, Unauthorized Exceptions are a 'HttpException' which is not logged by default. You can change which exceptions are logged using the `Exception.skipLog` & `Exception.log` config options.
